### ISSUE TITLE
Halt Device -- wrong condition code

### ISSUE THREAD

**Issue by WolfHimself** [rbowler/spinhawk#40]
When using a "Teletype" (ASCII terminal) to log in to VM/370, every console Prepare command ends with an attention signal.  Example:

    VM/370 ONLINE
    ?!

    .login maint
    !
    ENTER PASSWORD:
    .cpcmsSSS

    !
    DASD 190 LINKED R/W; R/O BY OPERATOR
    DASD 194 LINKED R/W; R/O BY OPERATOR
    LOGON AT 20:36:41 GMT FRIDAY 07/10/15
    !
    .q n

    !
    OPERATOR - 01F, MAINT    - 0F1
    !
    .i cms

    !
    RELEASE 6 CMS 12/25/78
    !
    .

    !
    .#cp log

    !
    CONNECT= 00:36:35 VIRTCPU= 000:00.02 TOTCPU= 000:00.22
    !
    LOGOFF AT 21:13:16 GMT FRIDAY 07/10/15

There are artifacts, here, from the implementation of commadpt.c but, besides that, VM prints an exclamation mark in response to Attention.  There are a lot of exclamation marks there.  IPL CMS never gets to the Ready message, because the VM read is retried over and over.  I used #CP to break out.

Prepare marks time when there is nothing else to do.  If the terminal user presses Break, this is Attention;  Prepare completes with CE+DE.  Before starting a new console task, VM uses Halt Device to stop the Prepare.  After that, Prepare is assumed to have completed without Attention.

The problem is that Halt Device signals, via dev->halt_device(dev), for the device to stop, then sets Condition Code 0, which indicates *failure*.  Normal, successful completion should be CC 1, with zeroes in the CSW status field.  Because Halt Device failed, VM (DMKCNS) takes the ensuing I/O interrupt as an Attention signal.

Quoting from the _S/370 PoP_ (09/74, p.202):

"_Condition Code 0_ indicates that HALT DEVICE cannot signal the control unit until an interrupt condition on the same subchannel is cleared.

"_Condition Code 1 with Control-Unit-Busy Status in the CSW_ indicates that HALT DEVICE cannot signal the control unit until the control-unit-end status is received from that control unit.

"_Condition Code 1 with Zeroes in the Status Field of the CSW_ indicates that the addressed device was selected and signaled to terminate the current operation, if any.

"_Condition Code 2_ indicates that the control unit cannot be signaled until the end of a busy condition in the channel.  The end of the busy condition can be detected by noting an interruption from the channel or by noting the results of repeatedly executing HALT DEVICE.

"_Condition Code 3_ indicates that manual intervention is required to allow HALT DEVICE to signal the control unit to terminate."

This can be cured in channel.c.  Here is the diff:

    361c361,364
    <             cc=0;                               /* @ISW */
    ---
    >             psa = (PSA_3XX*)( regs->mainstor + regs->PX );
    >             psa->csw[4] = 0;    /*  Store partial CSW       */
    >             psa->csw[5] = 0;
    >             cc = 1;             /*  Set CC for CSW stored   */

After applying it, the VM login looks like this:

    VM/370 ONLINE
    ?!

    .l maint

    ENTER PASSWORD:
    .cpcmsSSS

    DASD 190 LINKED R/W; R/O BY OPERATOR
    DASD 194 LINKED R/W; R/O BY OPERATOR
    LOGMSG - 22:05:10 GMT MONDAY 06/29/15
    * WELCOME TO THE FIRST LEVEL 3330 SYSTEM.
    LOGON AT 03:44:42 GMT SUNDAY 07/12/15
    .i cms

    RELEASE 6 CMS 12/25/78
    .

    R; T=0.01/0.01 03:44:48

    .l dmkcns(d

    FILENAME FILETYPE  FM  FORMAT    RECS BLOCKS   DATE   TIME
    DMKCNS   ASSEMBLE  A1  F    80  2181   219   2/22/15 14:46
    DMKCNS   UPDATES   A1  F    80    20     2   2/22/15 14:46
    DMKCNS   UPDLOG    A1  F    97   129    16   2/22/15 14:46
    R; T=0.01/0.01 03:44:59

    .log hold

    CONNECT= 00:00:44 VIRTCPU= 000:00.01 TOTCPU= 000:00.09
    LOGOFF AT 03:45:25 GMT SUNDAY 07/12/15

    VM/370 ONLINE

**Comment by Fish-Git2** [comment]
Hello WolfHimself,

I would like to investigate this issue to see whether or not it also exists in 4.0 Hyperion. I suspect it does and if so, I will create a separate github issue there for tracking purposes.

Can you explain how to recreate this? I have the VM/370 SixPack installed. What do I do? Add (a) 2703 BSC device(s)? What option(s) should I use? How do I connect to that device? Via ascii telnet? How do you do it?

Thanks!

Fish

**Comment by WolfHimself** [comment]
Fish:

If you have the SixPack...

You can get away with using the existing 2701 coded in DMKRIO at I/O addresses 020-02F.  Never mind that commadpt.c claims to be for a 2703.  DMKCNS treats the two adapters as if they were the same, anyway.  The important detail is the ADAPTER=TELE2 part.

Add a device to the Hercules configuration file like this:

    0020    2703 lport=2700 dial=IN lnctl=tele2 term=tty iskip=0a break=dumb

The "iskip" and "break" parameters are kludges to make the Telnet server implementation work.  If you will be using the Debian Telnet client, you will also need the parameter "skip=89c9ff".

This is a start/stop line.  I haven't bothered to analyze BSC.

The default SixPack enables the line at system IPL time.  Otherwise, the operator ENABLE command will do it.

Connect using the normal Telnet client.  Specify the port identified by "lport", 2700 in the example above.

Before you connect, here is the hardest part:

DMKCNS is broken in the FourPack, FivePack, and SixPack.  PTF UV02926 (APAR VM12126) inadvertently deleted two important source statements.  The result is that VM crashes and restarts when it gets Attention from a TTY.  This is not a problem if you install from Bob Abeles's starter system, which does not include the guilty PTF.  Here is a fix, in the form of local service:

The files are DMKCNS AUXLCL (for the FourPack system) and DMKCNS TTYFIX01.  DMKCNS AUXLCL contains:

    TTYFIX01 ABEND FRE005 WHEN ATTEMPTING TTY LOGIN

On the FivePack and SixPack systems, this line should be merged into DMKCNS AUXHRC instead.  I tested the base system only, not the AP and/or VR variants.

DMKCNS TTYFIX01 contains:

    ./ I 00640000 $ 640500
             B     CNSATTN        FIGURE OUT ATTENTION HANDLING    TTYFIX01
    ./ I 00647000 $ 647500
             B     CNSTART        TRY AGAIN                        TTYFIX01

After applying this service and generating a new system, you can actually log in via Telnet.

Let me know how it goes.

-- WolfHimself

**Comment by Fish-Git** [comment]
WolfHimself wrote:

> Fish:
> 
> If you have the SixPack...
> 
> You can get away with using the existing 2701 coded
> in DMKRIO at I/O addresses 020-02F. Never mind that
> commadpt.c claims to be for a 2703. [...]
<snip further details>

THANK YOU!

With your instructions I was able to confirm the problem does exist in 4.0 Hyperion as well. It has now been corrected however, per your patch. (See: https://github.com/hercules-390/hyperion/issues/59)

Thank you for reporting this issue and providing the patch!

Now all we need to do is wait for Roger to apply the same fix to Spinhawk. :)

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com
